Add email block renderer for Quote core block

yuliyan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

* This PR adds an email block renderer for the Quote core block and adds it to the allowed blocks in the email editor.

Closes [WOOPLUG-3864](https://linear.app/a8c/issue/WOOPLUG-3864/core-blocks-quote).

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Screenshot |
| ------ |
|  ![sHVMpCmlKhxRXfXN](https://github.com/user-attachments/assets/2724e97e-7c7e-416c-9ba0-b959e649d73d)  |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the block-based email editor from WooCommerce → Settings → Advanced → Features → **Block Email Editor (alpha)**
2. Open an email in the editor WooCommerce → Settings → **Emails**
3. Confirm the Quote block is present in the blocks list.
4. Add a Quote block to the email content and apply some customizations from the sidebar.
5. Send a test email from the Preview dropdown and confirm the applied styling for the Quote block is displayed correctly in the email.
6. Confirm integration tests for the Quote block pass.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Add an email block renderer for the Quote core block in the email block-based editor.</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

yuliyan committed:
Add email block renderer for Quote core block

yuliyan committed:
Add changelog entry

yuliyan committed:
Clean up redundant block styles

lysyjan reviewed (changes requested):
Hey @yuliyan 👋 
I tested your PR and I found a couple of issues:
1. There is a missing line on the left side of the block. It's missing in the rendered HTML and also in the email editor. You can check how it should look like in the post editor. Here is an example in the second image below:
<img width="707" alt="Screenshot 2025-04-17 at 16 59 25" src="https://github.com/user-attachments/assets/07f60745-c42e-4889-be0a-c2043b9102cb" />

---

<img width="800" alt="Screenshot 2025-04-17 at 17 00 10" src="https://github.com/user-attachments/assets/3744cf3e-7630-4264-b751-60af2404210b" />

2. The citation is not rendered in the email:
<img width="746" alt="Screenshot 2025-04-17 at 17 03 32" src="https://github.com/user-attachments/assets/19de6709-23b7-4a81-9c4c-40d5345ba318" />

3. Alignment for the quote block does not work in the template mode:
<img width="717" alt="Screenshot 2025-04-17 at 17 06 01" src="https://github.com/user-attachments/assets/206aeb0c-c950-400b-bf1b-858e784358aa" />

4. The selected alignment is not reflected in the rendered email.

5. There is a redundant margin in the email editor:
<img width="714" alt="Screenshot 2025-04-17 at 17 11 35" src="https://github.com/user-attachments/assets/392249b2-b7b7-4569-9093-5bd14993080e" />


Feel free to ping me if you need help addressing these issues or getting more details. I haven't tested the block across different email clients yet.

yuliyan committed:
Add citation content

yuliyan committed:
Reset quote block inline margin

yuliyan committed:
Disable alignment controls support for the Quote block

yuliyan committed:
Set the text-align attribute to the Quote wrapper

yuliyan commented:
Thank you for the detailed feedback, @lysyjan!

1. The border and padding styles for the Quote block in the post editor appear to be set in the theme. However, we don't seem to load the theme styles in the Email editor and thus if not set as an attribute, the border is missing in the preview and the rendered email markup. If we want to have a border by default, I suppose we can override the default values for the border attribute, but I don't think it'd be a good approach. What do you think?

2. I've added the citation content to the renderer. One thing that concerns me is that I'm not escaping the HTML, since the citation supports the typography controls (bold, italic, highlights, etc.) and the tags need to be preserved. I noticed we're handling it more or less the same way in other places (like the paragraph block, for example). Do you have any suggestions here?

3. I decided to remove the alignment controls in this case, since it seemed like this type of alignment (float) doesn't make much sense in the email markup context and wouldn't work as expected. Also, none of the other blocks seem to support this type of alignment. Let me know if I'm missing something.

4. Since I've removed the alignment controls for the Quote block, there's no need to add alignment support in the renderer. However, I noticed that the renderer was also missing logic to handle the Quote block's text alignment attribute, so I added that in. That said, there's another issue: while the citation aligns correctly, some child blocks, like paragraphs, set their own `text-align` attributes, which can override the alignment set by the Quote block. Do you think we should disable the text align controls as well, or see if it's possible to pass the text alignment value from the Quote block down to the children?

5. I've reset the inline margins for the `blockquote` element in the Email editor styles.

lysyjan commented:
> The border and padding styles for the Quote block in the post editor appear to be set in the theme. However, we don't seem to load the theme styles in the Email editor and thus if not set as an attribute, the border is missing in the preview and the rendered email markup. If we want to have a border by default, I suppose we can override the default values for the border attribute, but I don't think it'd be a good approach. What do you think?

@yuliyan I was investigating a possible approach with styles here, and I would extend [content-editor](https://github.com/woocommerce/woocommerce/blob/88f1c3c00c9bfc7c51f0b250623c4be761381d52/packages/php/email-editor/src/Engine/content-editor.css#L69-L91) styles. There are already some styles for the image block. In my opinion, the quote block renderer should also handle those styles for the correct look in emails.

I know it can be overwhelming, but we want to ensure the look in the editor and all supported email clients is almost identical.

lysyjan commented:
@yuliyan, I was testing how text alignment works in the post editor, and I would recommend the same. For example, when the text alignment is set to right, the default border is changed to right and when it is centered, there is no border.

I was also thinking about a possible conflict with the child blocks, and I would create a preprocessor for that. You can check how similar classes look in the PHP package and `Automattic\WooCommerce\EmailEditor\Engine\Renderer\ContentRenderer` namespace

yuliyan committed:
Add default border styles for the quote block

yuliyan committed:
Fix lint errors

yuliyan committed:
Add preprocessor to handle passing the textAlgin attribute to quote nested blocks

yuliyan committed:
Add changelog entry for @woocommerce/email-editor

yuliyan commented:
@lysyjan, thank you for the suggestions. I've [added the default border styles in the email editor](https://github.com/woocommerce/woocommerce/pull/57280/commits/8c153a52f6a8d2a89c25108d325365f0144b5e82) and the renderer and [created a new preprocessor to handle the text alignment](https://github.com/woocommerce/woocommerce/pull/57280/commits/a29ea8de129454bd9ad6ac9532958773fd7d428c) for the nested blocks. Could you take a look?

yuliyan committed:
Fix lint errors

yuliyan committed:
Move quote block styles to Integration/Core theme.json

yuliyan committed:
Disable core/quote styles in email-editor

yuliyan committed:
Fix Quote block styles for CSS inliner

yuliyan committed:
Handle Quote's typography in preprocessor

yuliyan committed:
Fix integration tests

yuliyan committed:
Add post processor for border styles

yuliyan committed:
Fix unit tests

yuliyan committed:
Revert unintended formatting changes

yuliyan committed:
Use <blockquote> for the wrapper and ensure email clients compatibility

yuliyan committed:
Fix tests

yuliyan commented:
@lysyjan, I've addressed the items that we discussed:

* Moved the default border styles to `Integration/Core/theme.json`.
* Created new post processor to handle properly setting the `border-style` declaration only where needed to properly render the expected borders across email clients.
* Removed the style variations for the Quote block (Plain, default) like we do for other blocks.
* Use `<blockquote>` for the wrapper, and `<cite>` for the citation content.

Could you take another look when you get the chance?

lysyjan commented on packages/php/email-editor/src/Integrations/Core/theme.json:
```diff
@@ -1,27 +1,62 @@
+				},
+				"spacing": {
+					"margin": {
+						"top": "0",
```
We don't want to reset the margin top and bottom because there is a default block spacing. This spacing can also be set in the layout settings. You can notice that in the current email editor, there is no vertical space between quote blocks, but when the email is rendered, there are spaces, which is correct.

yuliyan replied:
This was an unintentional change. I've removed the top and bottom margin resets.

lysyjan commented on packages/php/email-editor/tests/integration/Integrations/Core/Renderer/Blocks/Quote_Test.php:
```diff
@@ -0,0 +1,153 @@
+		$parsed_quote['attrs'] = array(
+			'backgroundColor' => '#abcdef',
+			'style'           => array(
+				'border'  => array(
+					'bottom' => array(
+						'color' => '#111111',
+						'width' => '1px',
+						'style' => 'dotted',
+					),
+					'left'   => array(
+						'color' => '#222222',
+						'width' => '2px',
+					),
+					'right'  => array(
+						'color' => '#333333',
+						'width' => '3px',
+					),
+					'top'    => array(
+						'color' => '#444444',
+						'width' => '4px',
+					),
+					'radius' => array(
+						'bottomLeft'  => '5px',
+						'bottomRight' => '10px',
+						'topLeft'     => '15px',
+						'topRight'    => '20px',
+					),
```
I am not sure about this test block configuration because the quote block does not allow configuration for separated border parts to be set up. It is always set for the whole block.

lysyjan replied:
Also, when I tested your changes, I had some trouble with the border color. I think the reason is that the border color is stored at another level. From what I checked, the color is stored under the key `borderColor`.

yuliyan replied:
I might be misunderstanding your comment, but from what I've seen both border configuration formats are supported. When a uniform border style is used, the block editor generates attributes in this format:

```php
'border' => array(
	'width' => ...,
	'style' => ...,
	// etc.
)
```

For mixed border styles, the attributes are parsed into this structure:

```php
'border' => array(
	'bottom' => array(
		'color' => ...,
		'width' => ...,
		'style' => ...,
	),
	'left' => ...,
	'right' => ...,
	'top' => ...,
)
```

As for the border color - you are correct. In the case of having uniform border styles, it is being stored as a top level attribute with `borderColor` key, and when having mixed styles, it's stored individually as a child of the respective border side. I had it handled at some point, but it seems like I forgot to transfer it after switching to the post processor approach.

I've now re-added it and updated the unit test to cover the case.

lysyjan commented on packages/php/email-editor/src/Integrations/Core/Renderer/Blocks/class-quote.php:
```diff
@@ -0,0 +1,134 @@
+
+		return sprintf(
+			'<!--[if mso | IE]><table align="left" role="presentation" border="0" cellpadding="0" cellspacing="0" width="100%%" style="%1$s"><tr><td style="%2$s" width="100%%"><![endif]-->
+			<blockquote class="email-block-quote %3$s" style="%1$s %2$s">
```
Thank you for adding the blockquote tag here. Unfortunately, when I tested the latest changes, I had some issues with the Gmail mobile app on EmailOnAcid. It looks like styling this HTML tag is not possible, and the left and right borders are missing. I am afraid we need to avoid using this tag because of the Gmail app.
<img width="398" alt="Screenshot 2025-05-06 at 18 51 56" src="https://github.com/user-attachments/assets/ea7738f7-61ae-4046-8c30-14c05145b049" />

yuliyan replied:
Apologies - I didn't realize it was broken. I updated the email source code using the `<blockquote>` tag on Email On Acid, but I was apparently looking at outdated screenshots and thought everything was working fine.

It turns out the `<blockquote>` tag also caused the border style issue you mentioned in your other comment. Reverting back to using tables seems to have resolved all of these issues.

Do you think we should also replace the `<cite>` tag with a table? From my testing, the `<cite>` tag appears to render well across email clients, but I wanted to double check.

lysyjan replied:
I think the `<cite>` tag is fine. I checked, and there are no default styles for that.

lysyjan commented on packages/php/email-editor/src/Engine/Renderer/ContentRenderer/Postprocessors/class-border-style-postprocessor.php:
```diff
@@ -0,0 +1,236 @@
+namespace Automattic\WooCommerce\EmailEditor\Engine\Renderer\ContentRenderer\Postprocessors;
+
+/**
+ * Ensures every border with border-width > 0 has a border-style, defaulting to 'solid' if missing, and removes extra border-style properties.
```
Good job with this class. It is pretty complex. Please can you extend the class description about the purpose and add some examples?

yuliyan replied:
Added a more detailed description about the issue that this post processor resolves.

lysyjan reviewed (commented):
The rendering is improved compared with the previous version. However, I still found some issues with EmailOnAcid. For example, Outlook clients are not well supported.

Here is one example:
Editor:
<img width="706" alt="Screenshot 2025-05-06 at 18 52 13" src="https://github.com/user-attachments/assets/b0fe4bdc-c786-48e3-b5da-1cd499b6fb05" />

Outlook:
<img width="708" alt="Screenshot 2025-05-06 at 18 51 37" src="https://github.com/user-attachments/assets/9f21cfc1-bfdd-4f27-972e-80d42958f481" />

I also noticed a problem with the spacing between blocks that is mentioned in another comment.

yuliyan committed:
Revert to using tables instead of <blockquote>

yuliyan committed:
Handle 'borderColor' attribute for uniform border styles

yuliyan committed:
Add more detailed description for the border style postprocessor

yuliyan committed:
Fix lint errors

lysyjan reviewed (approved):
Thank you for all your effort and patience.
The new quote block works nice now.

I tested it on my local with the latest WP 6.8.
I also used EmailOnAcid to check the output for different email clients. Here are some screenshots:
<img width="1812" alt="Screenshot 2025-05-07 at 17 47 57" src="https://github.com/user-attachments/assets/08b3bc54-6a2f-47f7-a060-dd765dd4f9e1" />
<img width="1061" alt="Screenshot 2025-05-07 at 17 47 39" src="https://github.com/user-attachments/assets/9379389a-5840-4a21-95ea-639ebe85ce84" />
<img width="996" alt="Screenshot 2025-05-07 at 17 47 04" src="https://github.com/user-attachments/assets/81603ebd-1f9d-400b-a346-82d104fbb9aa" />
<img width="644" alt="Screenshot 2025-05-07 at 17 46 17" src="https://github.com/user-attachments/assets/b4238761-3351-4128-905a-eb3bf4e0a022" />
